File: scripts/schema_lock.py
#!/usr/bin/env python3
"""Validate and lock paired English/Chinese TIDAS schema assets."""

from __future__ import annotations

import argparse
import hashlib
import json
import sys
from pathlib import Path
from typing import Any
from urllib.parse import unquote

from jsonschema import Draft7Validator, SchemaError

REPO_ROOT = Path(__file__).resolve().parents[1]
DEFAULT_SCHEMA_ROOT = REPO_ROOT / "src" / "tidas_tools" / "tidas"
LOCK_FILENAME = "schema.lock.json"
SCHEMA_SETS = {
    "en": "schemas",
    "zh": "schemas_zh",
}
DEFAULT_LOCALIZED_KEYS = ("description",)


class SchemaLockError(Exception):
    def __init__(self, messages: list[str]):
        self.messages = messages
        super().__init__("\n".join(messages))


def canonical_bytes(value: Any) -> bytes:
    return json.dumps(
        value,
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
    ).encode("utf-8")


def sha256_bytes(value: bytes) -> str:
    return hashlib.sha256(value).hexdigest()


def sha256_json(value: Any) -> str:
    return sha256_bytes(canonical_bytes(value))


def repo_relative(path: Path) -> str:
    try:
        return path.resolve().relative_to(REPO_ROOT).as_posix()
    except ValueError:
        return path.as_posix()


def read_json_file(path: Path) -> Any:
    return json.loads(path.read_text(encoding="utf-8"))


def contract_node(node: Any, localized_keys: set[str]) -> Any:
    if isinstance(node, dict):
        return {
            key: contract_node(value, localized_keys)
            for key, value in node.items()
            if key not in localized_keys
        }
    if isinstance(node, list):
        return [contract_node(value, localized_keys) for value in node]
    return node


def iter_json_refs(node: Any, path: str = "$"):
    if isinstance(node, dict):
        for key, value in node.items():
            child_path = f"{path}/{key}"
            if key == "$ref" and isinstance(value, str):
                yield child_path, value
            else:
                yield from iter_json_refs(value, child_path)
    elif isinstance(node, list):
        for index, value in enumerate(node):
            yield from iter_json_refs(value, f"{path}/{index}")


def decode_json_pointer(pointer: str) -> list[str]:
    if pointer == "":
        return []
    if not pointer.startswith("/"):
        raise ValueError(f"JSON pointer must start with '/': {pointer}")
    return [
        unquote(token).replace("~1", "/").replace("~0", "~")
        for token in pointer[1:].split("/")
    ]


def pointer_exists(document: Any, pointer: str) -> bool:
    current = document
    for token in decode_json_pointer(pointer):
        if isinstance(current, dict):
            if token not in current:
                return False
            current = current[token]
        elif isinstance(current, list):
            try:
                index = int(token)
            except ValueError:
                return False
            if index < 0 or index >= len(current):
                return False
            current = current[index]
        else:
            return False
    return True


def validate_local_refs(
    docs: dict[str, Any],
    schema_dir: Path,
    lang: str,
    errors: list[str],
) -> None:
    for current_name, document in sorted(docs.items()):
        for ref_path, ref in iter_json_refs(document):
            if ref.startswith(("http://", "https://")):
                continue

            if "#" in ref:
                file_part, pointer = ref.split("#", 1)
            else:
                file_part, pointer = ref, ""

            target_ref = Path(file_part) if file_part else Path(current_name)
            if target_ref.is_absolute() or ".." in target_ref.parts:
                errors.append(
                    f"{lang}:{current_name}:{ref_path}: unsupported local $ref {ref!r}"
                )
                continue

            target_name = target_ref.name
            target_path = schema_dir / target_ref
            if target_name not in docs:
                errors.append(
                    f"{lang}:{current_name}:{ref_path}: missing $ref target "
                    f"{repo_relative(target_path)}"
                )
                continue

            try:
                if not pointer_exists(docs[target_name], pointer):
                    errors.append(
                        f"{lang}:{current_name}:{ref_path}: missing JSON pointer "
                        f"#{pointer} in {target_name}"
                    )
            except ValueError as exc:
                errors.append(f"{lang}:{current_name}:{ref_path}: {exc}")


def collect_schema_set(
    schema_root: Path,
    lang: str,
    subdir: str,
    errors: list[str],
) -> dict[str, Any]:
    schema_dir = schema_root / subdir
    docs: dict[str, Any] = {}
    files: dict[str, dict[str, str]] = {}

    if not schema_dir.is_dir():
        errors.append(f"{lang}: missing schema directory {repo_relative(schema_dir)}")
        return {"docs": docs, "files": files}

    for path in sorted(schema_dir.glob("*.json")):
        try:
            document = read_json_file(path)
        except Exception as exc:  # noqa: BLE001 - include parse details in CI output.
            errors.append(f"{repo_relative(path)}: invalid JSON: {exc}")
            continue

        try:
            Draft7Validator.check_schema(document)
        except SchemaError as exc:
            errors.append(
                f"{repo_relative(path)}: invalid Draft 7 schema: {exc.message}"
            )

        docs[path.name] = document
        files[path.name] = {
            "contentSha256": sha256_bytes(path.read_bytes()),
        }

    return {"docs": docs, "files": files, "path": schema_dir}


def first_diffs(left: Any, right: Any, path: str = "$", limit: int = 12) -> list[str]:
    diffs: list[str] = []

    def walk(a: Any, b: Any, current_path: str) -> None:
        if len(diffs) >= limit:
            return
        if type(a) is not type(b):
            diffs.append(
                f"{current_path}: type {type(a).__name__} != {type(b).__name__}"
            )
            return
        if isinstance(a, dict):
            left_keys = set(a)
            right_keys = set(b)
            for key in sorted(left_keys - right_keys):
                if len(diffs) >= limit:
                    return
                diffs.append(f"{current_path}/{key}: only in en")
            for key in sorted(right_keys - left_keys):
                if len(diffs) >= limit:
                    return
                diffs.append(f"{current_path}/{key}: only in zh")
            for key in sorted(left_keys & right_keys):
                walk(a[key], b[key], f"{current_path}/{key}")
            return
        if isinstance(a, list):
            if len(a) != len(b):
                diffs.append(f"{current_path}: list length {len(a)} != {len(b)}")
            for index, (left_item, right_item) in enumerate(zip(a, b)):
                walk(left_item, right_item, f"{current_path}/{index}")
            return
        if a != b:
            diffs.append(f"{current_path}: {a!r} != {b!r}")

    walk(left, right, path)
    return diffs


def schema_root_label(schema_root: Path) -> str:
    try:
        return schema_root.resolve().relative_to(REPO_ROOT).as_posix()
    except ValueError:
        return schema_root.as_posix()


def aggregate_file_hash(files: dict[str, dict[str, str]], key: str) -> str:
    return sha256_json({name: values[key] for name, values in sorted(files.items())})


def build_lock(
    schema_root: Path = DEFAULT_SCHEMA_ROOT,
    localized_keys: tuple[str, ...] = DEFAULT_LOCALIZED_KEYS,
) -> dict[str, Any]:
    schema_root = schema_root.resolve()
    errors: list[str] = []
    localized_key_set = set(localized_keys)
    collected: dict[str, dict[str, Any]] = {}

    for lang, subdir in SCHEMA_SETS.items():
        collected[lang] = collect_schema_set(schema_root, lang, subdir, errors)

    en_names = set(collected["en"]["docs"])
    zh_names = set(collected["zh"]["docs"])
    for name in sorted(en_names - zh_names):
        errors.append(f"schema file missing from zh set: {name}")
    for name in sorted(zh_names - en_names):
        errors.append(f"schema file missing from en set: {name}")

    for lang, result in collected.items():
        validate_local_refs(result["docs"], result["path"], lang, errors)

    shared_names = sorted(en_names & zh_names)
    pair_files: dict[str, dict[str, str]] = {}

    for name in shared_names:
        en_contract = contract_node(collected["en"]["docs"][name], localized_key_set)
        zh_contract = contract_node(collected["zh"]["docs"][name], localized_key_set)
        en_contract_hash = sha256_json(en_contract)
        zh_contract_hash = sha256_json(zh_contract)

        collected["en"]["files"][name]["contractSha256"] = en_contract_hash
        collected["zh"]["files"][name]["contractSha256"] = zh_contract_hash

        if en_contract != zh_contract:
            diffs = "\n    ".join(first_diffs(en_contract, zh_contract))
            errors.append(
                f"{name}: en/zh contract differs after removing localized keys "
                f"{sorted(localized_keys)}:\n    {diffs}"
            )

        pair_files[name] = {
            "contractSha256": en_contract_hash,
            "enContentSha256": collected["en"]["files"][name]["contentSha256"],
            "zhContentSha256": collected["zh"]["files"][name]["contentSha256"],
        }

    if errors:
        raise SchemaLockError(errors)

    schema_sets = {}
    for lang, subdir in SCHEMA_SETS.items():
        files = dict(sorted(collected[lang]["files"].items()))
        schema_sets[lang] = {
            "path": subdir,
            "fileCount": len(files),
            "contentAggregateSha256": aggregate_file_hash(files, "contentSha256"),
            "contractAggregateSha256": aggregate_file_hash(files, "contractSha256"),
            "files": files,
        }

    return {
        "version": 1,
        "schemaRoot": schema_root_label(schema_root),
        "allowedLocalizedKeys": list(localized_keys),
        "generation": {
            "tool": "scripts/schema_lock.py",
            "mode": "deterministic-v1",
        },
        "schemaSets": schema_sets,
        "translationPairs": {
            "fileCount": len(pair_files),
            "contractAggregateSha256": sha256_json(
                {
                    name: values["contractSha256"]
                    for name, values in sorted(pair_files.items())
                }
            ),
            "files": dict(sorted(pair_files.items())),
        },
    }


def lock_path(schema_root: Path) -> Path:
    return schema_root / LOCK_FILENAME


def dump_lock(lock: dict[str, Any]) -> str:
    return json.dumps(lock, ensure_ascii=False, indent=2, sort_keys=True) + "\n"


def update_lock(schema_root: Path) -> int:
    lock = build_lock(schema_root)
    target = lock_path(schema_root)
    target.write_text(dump_lock(lock), encoding="utf-8")
    print(f"Updated {repo_relative(target)}")
    return 0


def check_lock(schema_root: Path) -> int:
    expected = build_lock(schema_root)
    target = lock_path(schema_root)

    if not target.is_file():
        print(
            f"Missing {repo_relative(target)}. "
            "Run: uv run python scripts/schema_lock.py update",
            file=sys.stderr,
        )
        return 1

    try:
        actual = read_json_file(target)
    except Exception as exc:  # noqa: BLE001 - include parse details in CI output.
        print(f"{repo_relative(target)} is not valid JSON: {exc}", file=sys.stderr)
        return 1

    if actual != expected:
        print(
            f"{repo_relative(target)} is stale. "
            "Run: uv run python scripts/schema_lock.py update",
            file=sys.stderr,
        )
        for diff in first_diffs(actual, expected, limit=20):
            print(f"  {diff}", file=sys.stderr)
        return 1

    pair_count = expected["translationPairs"]["fileCount"]
    aggregate = expected["translationPairs"]["contractAggregateSha256"]
    print(f"Schema lock OK: {pair_count} schema pairs, contract {aggregate}")
    return 0


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "command",
        choices=("check", "update"),
        help="check the committed lock or update it from current schema assets",
    )
    parser.add_argument(
        "--schema-root",
        type=Path,
        default=DEFAULT_SCHEMA_ROOT,
        help="directory containing schemas/, schemas_zh/, and schema.lock.json",
    )
    return parser.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    args = parse_args(sys.argv[1:] if argv is None else argv)
    try:
        if args.command == "update":
            return update_lock(args.schema_root)
        return check_lock(args.schema_root)
    except SchemaLockError as exc:
        print("Schema lock validation failed:", file=sys.stderr)
        for message in exc.messages:
            print(f"  - {message}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/tidas-assets/src/schema_lock.rs
use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::{Component, Path, PathBuf};

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::{AssetError, sha256_hex};

pub const SCHEMA_LOCK_PATH: &str = "assets/tidas/schema.lock.json";
const SCHEMA_ROOT: &str = "assets/tidas";
const LOCALIZED_KEY: &str = "description";

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SchemaLockV1 {
    allowed_localized_keys: Vec<String>,
    generation: SchemaLockGeneration,
    schema_root: String,
    schema_sets: BTreeMap<String, SchemaSetLock>,
    translation_pairs: TranslationPairLock,
    version: u8,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
struct SchemaLockGeneration {
    mode: String,
    tool: String,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
struct SchemaSetLock {
    content_aggregate_sha256: String,
    contract_aggregate_sha256: String,
    file_count: usize,
    files: BTreeMap<String, SchemaFileLock>,
    path: String,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
struct SchemaFileLock {
    content_sha256: String,
    contract_sha256: String,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
struct TranslationPairLock {
    contract_aggregate_sha256: String,
    file_count: usize,
    files: BTreeMap<String, TranslationPairFileLock>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
#[allow(clippy::struct_field_names)] // Wire names preserve the historical lock schema.
struct TranslationPairFileLock {
    contract_sha256: String,
    en_content_sha256: String,
    zh_content_sha256: String,
}

struct CollectedSchemaSet {
    documents: BTreeMap<String, Value>,
    content_hashes: BTreeMap<String, String>,
}

pub fn schema_lock_from_filesystem(repo_root: &Path) -> Result<SchemaLockV1, AssetError> {
    let schema_root = repo_root.join(SCHEMA_ROOT);
    let en = collect_schema_set(&schema_root.join("schemas"))?;
    let zh = collect_schema_set(&schema_root.join("schemas_zh"))?;

    let en_names: BTreeSet<&str> = en.documents.keys().map(String::as_str).collect();
    let zh_names: BTreeSet<&str> = zh.documents.keys().map(String::as_str).collect();
    if en_names != zh_names {
        return Err(AssetError::SchemaLock(format!(
            "English and Chinese schema file sets differ; missing_zh={:?}, missing_en={:?}",
            en_names.difference(&zh_names).collect::<Vec<_>>(),
            zh_names.difference(&en_names).collect::<Vec<_>>()
        )));
    }

    validate_local_refs(&en.documents, "en")?;
    validate_local_refs(&zh.documents, "zh")?;

    let mut en_files = BTreeMap::new();
    let mut zh_files = BTreeMap::new();
    let mut pair_files = BTreeMap::new();
    for name in en.documents.keys() {
        let en_contract = contract_node(
            en.documents
                .get(name)
                .expect("schema name came from the same map"),
        );
        let zh_contract = contract_node(
            zh.documents
                .get(name)
                .expect("schema file sets were verified equal"),
        );
        if en_contract != zh_contract {
            return Err(AssetError::SchemaLock(format!(
                "{name}: English and Chinese contracts differ after removing `{LOCALIZED_KEY}`"
            )));
        }
        let contract_sha256 = sha256_json(&en_contract)?;
        let en_content_sha256 = en
            .content_hashes
            .get(name)
            .expect("every parsed document has a content hash")
            .clone();
        let zh_content_sha256 = zh
            .content_hashes
            .get(name)
            .expect("every parsed document has a content hash")
            .clone();
        en_files.insert(
            name.clone(),
            SchemaFileLock {
                content_sha256: en_content_sha256.clone(),
                contract_sha256: contract_sha256.clone(),
            },
        );
        zh_files.insert(
            name.clone(),
            SchemaFileLock {
                content_sha256: zh_content_sha256.clone(),
                contract_sha256: contract_sha256.clone(),
            },
        );
        pair_files.insert(
            name.clone(),
            TranslationPairFileLock {
                contract_sha256,
                en_content_sha256,
                zh_content_sha256,
            },
        );
    }

    let mut schema_sets = BTreeMap::new();
    schema_sets.insert("en".to_owned(), schema_set_lock("schemas", en_files)?);
    schema_sets.insert("zh".to_owned(), schema_set_lock("schemas_zh", zh_files)?);
    let pair_contracts: BTreeMap<&str, &str> = pair_files
        .iter()
        .map(|(name, lock)| (name.as_str(), lock.contract_sha256.as_str()))
        .collect();

    Ok(SchemaLockV1 {
        allowed_localized_keys: vec![LOCALIZED_KEY.to_owned()],
        generation: SchemaLockGeneration {
            mode: "deterministic-v1".to_owned(),
            tool: "tidas-asset-lock".to_owned(),
        },
        schema_root: SCHEMA_ROOT.to_owned(),
        schema_sets,
        translation_pairs: TranslationPairLock {
            contract_aggregate_sha256: sha256_json(&pair_contracts)?,
            file_count: pair_files.len(),
            files: pair_files,
        },
        version: 1,
    })
}

pub fn write_schema_lock(repo_root: &Path) -> Result<PathBuf, AssetError> {
    let lock = schema_lock_from_filesystem(repo_root)?;
    let output_path = repo_root.join(SCHEMA_LOCK_PATH);
    let mut bytes = serde_json::to_vec_pretty(&lock)?;
    bytes.push(b'\n');
    fs::write(&output_path, bytes)?;
    Ok(output_path)
}

pub fn check_filesystem_schema_lock(repo_root: &Path) -> Result<(), AssetError> {
    let expected = schema_lock_from_filesystem(repo_root)?;
    let actual: SchemaLockV1 =
        serde_json::from_slice(&fs::read(repo_root.join(SCHEMA_LOCK_PATH))?)?;
    if actual == expected {
        Ok(())
    } else {
        Err(AssetError::SchemaLock(format!(
            "{SCHEMA_LOCK_PATH} is stale; run `cargo run -p tidas-assets --bin tidas-asset-lock -- write`"
        )))
    }
}

fn collect_schema_set(schema_dir: &Path) -> Result<CollectedSchemaSet, AssetError> {
    if !schema_dir.is_dir() {
        return Err(AssetError::SchemaLock(format!(
            "missing schema directory {}",
            schema_dir.display()
        )));
    }
    let mut paths = fs::read_dir(schema_dir)?
        .map(|entry| entry.map(|item| item.path()))
        .collect::<Result<Vec<_>, _>>()?;
    paths.retain(|path| path.extension().and_then(|value| value.to_str()) == Some("json"));
    paths.sort();

    let mut documents = BTreeMap::new();
    let mut content_hashes = BTreeMap::new();
    for path in paths {
        let name = path
            .file_name()
            .and_then(|value| value.to_str())
            .ok_or_else(|| {
                AssetError::SchemaLock(format!("non-UTF-8 schema path {}", path.display()))
            })?
            .to_owned();
        let bytes = fs::read(&path)?;
        let document: Value = serde_json::from_slice(&bytes)?;
        jsonschema::draft7::meta::validate(&document).map_err(|error| {
            AssetError::SchemaLock(format!(
                "{} is not a valid Draft 7 schema: {error}",
                path.display()
            ))
        })?;
        content_hashes.insert(name.clone(), sha256_hex(&normalize_line_endings(&bytes)));
        documents.insert(name, document);
    }
    Ok(CollectedSchemaSet {
        documents,
        content_hashes,
    })
}

fn validate_local_refs(
    documents: &BTreeMap<String, Value>,
    language: &str,
) -> Result<(), AssetError> {
    for (current_name, document) in documents {
        visit_refs(document, &mut |reference| {
            if reference.starts_with("http://") || reference.starts_with("https://") {
                return Ok(());
            }
            let (file_part, pointer) = reference
                .split_once('#')
                .map_or((reference, ""), |(file, pointer)| (file, pointer));
            let target = if file_part.is_empty() {
                Path::new(current_name)
            } else {
                Path::new(file_part)
            };
            if target.is_absolute()
                || target
                    .components()
                    .any(|component| component == Component::ParentDir)
            {
                return Err(AssetError::SchemaLock(format!(
                    "{language}:{current_name}: unsupported local reference `{reference}`"
                )));
            }
            let target_name = target
                .file_name()
                .and_then(|value| value.to_str())
                .ok_or_else(|| {
                    AssetError::SchemaLock(format!(
                        "{language}:{current_name}: invalid local reference `{reference}`"
                    ))
                })?;
            let target_document = documents.get(target_name).ok_or_else(|| {
                AssetError::SchemaLock(format!(
                    "{language}:{current_name}: missing local reference target `{target_name}`"
                ))
            })?;
            if !pointer_exists(target_document, pointer)? {
                return Err(AssetError::SchemaLock(format!(
                    "{language}:{current_name}: missing JSON pointer `#{pointer}` in `{target_name}`"
                )));
            }
            Ok(())
        })?;
    }
    Ok(())
}

fn visit_refs(
    value: &Value,
    visitor: &mut impl FnMut(&str) -> Result<(), AssetError>,
) -> Result<(), AssetError> {
    match value {
        Value::Object(object) => {
            for (key, child) in object {
                if key == "$ref" {
                    if let Some(reference) = child.as_str() {
                        visitor(reference)?;
                    }
                } else {
                    visit_refs(child, visitor)?;
                }
            }
        }
        Value::Array(items) => {
            for item in items {
                visit_refs(item, visitor)?;
            }
        }
        _ => {}
    }
    Ok(())
}

fn pointer_exists(document: &Value, pointer: &str) -> Result<bool, AssetError> {
    if pointer.is_empty() {
        return Ok(true);
    }
    if !pointer.starts_with('/') {
        return Err(AssetError::SchemaLock(format!(
            "JSON pointer must start with `/`: {pointer}"
        )));
    }
    let mut current = document;
    for raw_token in pointer[1..].split('/') {
        let token = decode_percent(raw_token)?
            .replace("~1", "/")
            .replace("~0", "~");
        match current {
            Value::Object(object) => {
                let Some(next) = object.get(&token) else {
                    return Ok(false);
                };
                current = next;
            }
            Value::Array(items) => {
                let Ok(index) = token.parse::<usize>() else {
                    return Ok(false);
                };
                let Some(next) = items.get(index) else {
                    return Ok(false);
                };
                current = next;
            }
            _ => return Ok(false),
        }
    }
    Ok(true)
}

fn decode_percent(value: &str) -> Result<String, AssetError> {
    let input = value.as_bytes();
    let mut decoded = Vec::with_capacity(input.len());
    let mut index = 0;
    while index < input.len() {
        if input[index] == b'%' {
            let Some(high) = input.get(index + 1).and_then(|byte| hex_digit(*byte)) else {
                return Err(AssetError::SchemaLock(format!(
                    "invalid percent encoding in JSON pointer token `{value}`"
                )));
            };
            let Some(low) = input.get(index + 2).and_then(|byte| hex_digit(*byte)) else {
                return Err(AssetError::SchemaLock(format!(
                    "invalid percent encoding in JSON pointer token `{value}`"
                )));
            };
            decoded.push(high * 16 + low);
            index += 3;
        } else {
            decoded.push(input[index]);
            index += 1;
        }
    }
    String::from_utf8(decoded).map_err(|_| {
        AssetError::SchemaLock(format!(
            "percent-decoded JSON pointer token is not UTF-8: `{value}`"
        ))
    })
}

const fn hex_digit(value: u8) -> Option<u8> {
    match value {
        b'0'..=b'9' => Some(value - b'0'),
        b'a'..=b'f' => Some(value - b'a' + 10),
        b'A'..=b'F' => Some(value - b'A' + 10),
        _ => None,
    }
}

fn contract_node(value: &Value) -> Value {
    match value {
        Value::Object(object) => Value::Object(
            object
                .iter()
                .filter(|(key, _)| key.as_str() != LOCALIZED_KEY)
                .map(|(key, child)| (key.clone(), contract_node(child)))
                .collect(),
        ),
        Value::Array(items) => Value::Array(items.iter().map(contract_node).collect()),
        _ => value.clone(),
    }
}

fn schema_set_lock(
    path: &str,
    files: BTreeMap<String, SchemaFileLock>,
) -> Result<SchemaSetLock, AssetError> {
    let content_hashes: BTreeMap<&str, &str> = files
        .iter()
        .map(|(name, lock)| (name.as_str(), lock.content_sha256.as_str()))
        .collect();
    let contract_hashes: BTreeMap<&str, &str> = files
        .iter()
        .map(|(name, lock)| (name.as_str(), lock.contract_sha256.as_str()))
        .collect();
    Ok(SchemaSetLock {
        content_aggregate_sha256: sha256_json(&content_hashes)?,
        contract_aggregate_sha256: sha256_json(&contract_hashes)?,
        file_count: files.len(),
        files,
        path: path.to_owned(),
    })
}

fn sha256_json(value: &impl Serialize) -> Result<String, AssetError> {
    Ok(sha256_hex(&serde_json::to_vec(value)?))
}

fn normalize_line_endings(bytes: &[u8]) -> Vec<u8> {
    let mut normalized = Vec::with_capacity(bytes.len());
    let mut index = 0;
    while index < bytes.len() {
        if bytes[index] == b'\r' && bytes.get(index + 1) == Some(&b'\n') {
            normalized.push(b'\n');
            index += 2;
        } else {
            normalized.push(bytes[index]);
            index += 1;
        }
    }
    normalized
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn json_pointer_supports_objects_arrays_and_escapes() {
        let document = serde_json::json!({"a/b": [{"~key": true}], "with space": true});
        assert!(pointer_exists(&document, "/a~1b/0/~0key").unwrap());
        assert!(pointer_exists(&document, "/with%20space").unwrap());
        assert!(!pointer_exists(&document, "/a~1b/1").unwrap());
    }

    #[test]
    fn contract_nodes_drop_only_localized_descriptions() {
        let document = serde_json::json!({
            "description": "localized",
            "properties": {"description": {"type": "string"}},
            "title": "stable"
        });
        assert_eq!(
            contract_node(&document),
            serde_json::json!({"properties": {}, "title": "stable"})
        );
    }
}
